---
title: "Prep Fraction Dataframes"
author: "DCI Bioinformatics"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    code_folding: hide
    number_sections: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE, 
  comment = '',
  results = "asis",
  warning = FALSE, 
  message = FALSE, 
  fig.align = 'center',
  size = 'footnotesize',
  fig.asp = 0.5, 
  out.width = "80%",
  fig.width = 7,
  tidy.opts = list(keep.blank.line = FALSE,
                   width.cutoff = 70)
)
```

```{r dtsetup, include=FALSE}
# init step to make sure that DT dependencies are loaded
# if this is not pre-set, tables may not display properly from
# within for loops

htmltools::tagList(
  DT::datatable(
    iris, 
    rownames = FALSE,
    filter = "top",
    extensions = 'Buttons', 
    options = list(
      dom = 'Bfrtip',
      scrollX = TRUE,
      buttons = list(
        list(extend = 'colvis', 
             columns = c(0,1,2,3,4)),
        list(extend = "colvisGroup",
             text = "Show All",
             show = ":hidden")),
      columnDefs = list(
        list(visible = FALSE,
             targets = c(0,1))
      )
    )
  )
)

```

# Prerequisites

#### Container {-}

```{bash eval=F}

sif=/opt/apps/containers/community/ACC/dcibioinformaticsR-v2.3.sif
wd=/hpc/group/chsi/RhCMV
singularity shell --no-home --bind ${wd}:${wd}:rw ${sif}
R

```

#### Paths & Libraries {-}

```{r message=FALSE, warning=FALSE}
startdt <- date() # start time

library(tidyverse)
library(DT)
library(glue)

wd <- "/hpc/group/chsi/RhCMV"
source(file.path(wd, "Code", "config.R"))
source(file.path(wd, "Code","functions","utilities.R"))
source(file.path(wd, "Code","functions","strainSNP_utilities.R"))

# output directory for intermediates generated by this script
procdir.5fractiondfs

# directory with inputs generated by 2_nucdiff.sh
procdir.2nucdiff
gff.cols <- c("seqid","source","type","start","end",
              "score","strand","phase","attributes")

# directory with inputs generated by 4_summarize-mpileup
procdir.4summpileup


```

#### Functions {-}

```{r}
# Processing/Make base-level strain frequency dataframes
collect_plotdfs <- function(basesumm.ann, sSNPbases, 
                            curr.reference, alt.reference){
  
  # set up the x axis
  gg.pos <- sSNPbases %>%
    mutate(pos.chr = as.character(POS),
           pos = as.numeric(POS)) %>%
    distinct() %>%
    arrange(pos)
  gg.pos$pos.seq <- seq(1:nrow(gg.pos))
  
  # reformat the sample position frequencies wider
  basesumm.w <- basesumm.ann %>%
    pivot_wider(id_cols = pos,
                names_from = var.type,
                values_from = count,
                values_fill = 0) %>% 
    mutate(coverage = ref.base + alt.base + other,
           log2coverage = log2(coverage + 0.1),
           pos = as.numeric(pos))
  
  # join the sample position frequencies to the full set of strain SNPs
  plotdf <- gg.pos %>%
    left_join(basesumm.w, by = "pos") %>%
    mutate(ref.base = ifelse(is.na(ref.base), 0, ref.base),
           alt.base = ifelse(is.na(alt.base), 0, alt.base),
           other = ifelse(is.na(other), 0, other),
           coverage = ifelse(is.na(coverage), 0, coverage),
           log2coverage = ifelse(is.na(log2coverage), 0, log2coverage)) %>%
    mutate(ref.base.frac = ref.base/coverage,
           alt.base.frac = alt.base/coverage,
           other.frac = other/coverage)
  
  # make counts plotdf
  plotdf.counts <- plotdf %>%
    dplyr::select(pos, ref.base, alt.base, other) %>%
    pivot_longer(cols = c(ref.base, alt.base, other),
                 names_to = "var.type",
                 values_to = "count") %>%
    mutate(var.type.name = case_when(
      var.type == "ref.base" ~ curr.reference,
      var.type == "alt.base" ~ alt.reference,
      TRUE ~ var.type
    ))
  
  # make fraction plotdf
  plotdf.fracs <- plotdf %>%
    dplyr::select(pos, ref.base.frac, alt.base.frac, other.frac) %>%
    pivot_longer(cols = c(ref.base.frac, alt.base.frac, other.frac),
                 names_to = "var.type",
                 values_to = "frac") %>%
    mutate(var.type.name = case_when(
      var.type == "ref.base.frac" ~ curr.reference,
      var.type == "alt.base.frac" ~ alt.reference,
      TRUE ~ var.type
    ))
  
  l <- list(plotdf = plotdf,
            plotdf.counts = plotdf.counts,
            plotdf.fracs = plotdf.fracs)
  
  return(l)
  
}

lsf.str()

```

#### Formatting {-}

```{r message=FALSE, warning=FALSE}
# Table and Figure numbers
# for newlab(), see functions/utilities.R
figlab <- "Figure 0"
tablab <- "Table 0"

```

# Notes

Reference strains:

- VID965-FL(GCA_027929765.1)
- UCD52 (GCA_027930255.1)

Alignment types:

- *RvidQucd* = UCD52 (query) was aligned to VID965 (reference)
- *QvidRucd* = VID965 (query) was aligned to UCD52 (reference)

# Load data

## Load sample bam base frequencies

- Run `samtools mpileup` on all samples, see `1_get-bam-coverage-mpileup.sh`
- Base summaries were generated in `4_summarize-mpileup.R` (this step takes a very long time to run! 9 hours!)

```{r}
basesumm.vid <- readRDSobj(fdir = procdir.4summpileup, fname = "basesumm-lst-vid-n14.RDS")
basesumm.ucd <- readRDSobj(fdir = procdir.4summpileup, fname = "basesumm-lst-ucd-n14.RDS")

# list of dataframes, 1 per sample
names(basesumm.vid)

tablab <- newlab(tablab, caption = "First 6 rows of summarized mpileup results for 1 sample")
head(basesumm.vid$`KH14P14`) %>%
  knitr::kable(caption = tablab)

```

## Load nucdiff local differences and identify strain SNPs

- Local differences between the reference strains were identified using Nucdiff, see `2_nucdiff.sh`
- Local differences that are single nucleotide polymorphisms (SNPs) are selected using the custom function `get_strainSNPs` defined in `functions/strainSNP_utilities.R`

```{r}
## VID ##---------------------------------------------------##
sSNPbases.vid <- load_select_SNPbases(aln.type = "R_VID_Q_UCD", procdir.2nucdiff, gff.cols)
nrow(sSNPbases.vid) # 2862

## UCD ##---------------------------------------------------##
sSNPbases.ucd <- load_select_SNPbases(aln.type = "Q_VID_R_UCD", procdir.2nucdiff, gff.cols)
nrow(sSNPbases.ucd) # 2865

tablab <- newlab(tablab, caption = "First 6 rows of strain SNPs in UCD")
head(sSNPbases.ucd) %>%
  knitr::kable(caption = tablab)

```

# Make base-level strain frequency dataframes

### Select sample bam base frequencies at strain SNP positions

- For each sample, select the bam base frequencies (i.e. contained in `basesumm.vid`) that correspond to strain SNP positions (i.e., contained in `sSNPbases.vid`) and annotate the frequencies based on if the base matches the VID or UCD allele at the strain SNP position
- Save the results to `basesumm-ann-lst-<REF>-n14.RDS`

```{r}
preproc <- FALSE

## VID ##---------------------------------------------------##
sSNPbases <- sSNPbases.vid
outfile <- file.path(procdir.5fractiondfs,
                     "basesumm-ann-lst-vid-n14.RDS") 

if (preproc) {
  
  # identify reference alleles for each strain at each strain SNP position
  gg.pos <- sSNPbases %>%
    mutate(pos = as.character(POS))
  length(unique(gg.pos$pos)) # 2862 unique positions
  
  # filter to positions where strains differ and join allele info
  basesumm.ann.lst <- list()
  for (i in 1:length(basesumm.vid)) {
  
    basesumm.ann <- basesumm.vid[[i]] %>%
      filter(Var1 %in% c("A","T","C","G","N")) %>%
      filter(pos %in% gg.pos$pos) %>%
      left_join(gg.pos, by = c("pos")) %>%
      mutate(var.type = case_when(
        Var1 == REF ~ "ref.base",
        Var1 == ALT ~ "alt.base",
        Var1 != REF & Var1 != ALT ~ "other",
        TRUE ~ as.character(NA)
      )) %>%
      group_by(pos, var.type) %>%
      summarize(n = n(),
                count = sum(Freq))
    
    basesumm.ann.lst[[i]] <- basesumm.ann
  
  }
  
  names(basesumm.ann.lst) <- names(basesumm.vid)
  saveRDS(basesumm.ann.lst, file = outfile)
  
}

## UCD ##---------------------------------------------------##
sSNPbases <- sSNPbases.ucd
outfile <- file.path(procdir.5fractiondfs,
                     "basesumm-ann-lst-ucd-n14.RDS") 

if (preproc) {
  
  # identify reference alleles for each strain at each strain SNP position
  gg.pos <- sSNPbases %>%
    mutate(pos = as.character(POS))
  length(unique(gg.pos$pos)) # 2865 unique positions
  
  # filter to positions where strains differ and join allele info
  basesumm.ann.lst <- list()
  for (i in 1:length(basesumm.ucd)) {
  
    basesumm.ann <- basesumm.ucd[[i]] %>%
      filter(Var1 %in% c("A","T","C","G","N")) %>%
      filter(pos %in% gg.pos$pos) %>%
      left_join(gg.pos, by = c("pos")) %>%
      mutate(var.type = case_when(
        Var1 == REF ~ "ref.base",
        Var1 == ALT ~ "alt.base",
        Var1 != REF & Var1 != ALT ~ "other",
        TRUE ~ as.character(NA)
      )) %>%
      group_by(pos, var.type) %>%
      summarize(n = n(),
                count = sum(Freq))
    
    basesumm.ann.lst[[i]] <- basesumm.ann
  
  }
  
  names(basesumm.ann.lst) <- names(basesumm.ucd)
  saveRDS(basesumm.ann.lst, file = outfile)
  
}

```

### Reshape data into dataframes for downstream use

- Reshape the base-level strain frequency data (i.e.  `sample-strain-frequencies/basesumm-ann-lst-<REF>-n14.RDS`) into dataframes that combine samples using the custom function `collect_plotdfs` defined in this Rmd
- Three dataframe types are generated: `plotdf`, `plotdf.counts`, `plotdf.fracs`

```{r}
## VID ##---------------------------------------------------##
sSNPbases <- sSNPbases.vid
basesummannfile <- file.path(procdir.5fractiondfs,
                             "basesumm-ann-lst-vid-n14.RDS") 
basesumm.ann <- readRDS(basesummannfile)
curr.reference <- "VID965"
alt.reference <- "UCD52"

l.vid <- list()
for (i in 1:length(basesumm.ann)) {
  
 l <- collect_plotdfs(basesumm.ann[[i]], 
                      sSNPbases, 
                      curr.reference, 
                      alt.reference)
 
 l.vid[[i]] <- l
  
}

names(l.vid) <- names(basesumm.ann)
  

## UCD ##---------------------------------------------------##
sSNPbases <- sSNPbases.ucd
basesummannfile <- file.path(procdir.5fractiondfs,
                             "basesumm-ann-lst-ucd-n14.RDS") 
basesumm.ann <- readRDS(basesummannfile)
alt.reference <- "VID965"
curr.reference <- "UCD52"

l.ucd <- list()
for (i in 1:length(basesumm.ann)) {
  
 l <- collect_plotdfs(basesumm.ann[[i]], 
                      sSNPbases, 
                      curr.reference, 
                      alt.reference)
 
 l.ucd[[i]] <- l
  
}

names(l.ucd) <- names(basesumm.ann)
  

tablab <- newlab(tablab, caption = "First 10 rows of 'plotdf' for 1 sample aligned to UCD")
head(l.ucd$`KH14P14`$plotdf) %>%
  simple.DT(title = tablab) %>%
  DT::formatRound(digits = 2, columns = "log2coverage")

# Is the the `plotdf.counts` data frame still needed?
tablab <- newlab(tablab, caption = "First 10 rows of 'plotdf.counts' for 1 sample aligned to UCD")
head(l.ucd$`KH14P14`$plotdf.counts) %>%
  simple.DT(title = tablab)

tablab <- newlab(tablab, caption = "First 10 rows of 'plotdf.fracs' for 1 sample aligned to UCD")
head(l.ucd$`KH14P14`$plotdf.fracs) %>%
  simple.DT(title = tablab)

```

# Save dataframes

```{r}
saveRDSobj(l.vid, outdir = procdir.5fractiondfs)
saveRDSobj(l.ucd, outdir = procdir.5fractiondfs)

```

# Session information

```{r, results='hold'}
sessionInfo()

print(glue("Start Time: ",startdt))
print(glue("End Time: ",date()))

```
