---
title: "Filtering SNVs"
author: "DCI Bioinformatics"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    code_folding: hide
    number_sections: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE, 
  comment = '',
  results = "asis",
  warning = FALSE, 
  message = FALSE, 
  fig.align = 'center',
  size = 'footnotesize',
  fig.asp = 0.5, 
  out.width = "80%",
  fig.width = 7,
  tidy.opts = list(keep.blank.line = FALSE,
                   width.cutoff = 70)
)
```

```{r dtsetup, include=FALSE}
# init step to make sure that DT dependencies are loaded
# if this is not pre-set, tables may not display properly from
# within for loops

htmltools::tagList(
  DT::datatable(
    iris, 
    rownames = FALSE,
    filter = "top",
    extensions = 'Buttons', 
    options = list(
      dom = 'Bfrtip',
      scrollX = TRUE,
      buttons = list(
        list(extend = 'colvis', 
             columns = c(0,1,2,3,4)),
        list(extend = "colvisGroup",
             text = "Show All",
             show = ":hidden")),
      columnDefs = list(
        list(visible = FALSE,
             targets = c(0,1))
      )
    )
  )
)

```

# Prerequisites

#### Container {-}

```{bash eval=F}

# for R chunks
sif=/opt/apps/containers/community/ACC/dcibioinformaticsR-v2.3.sif
wd=/hpc/group/chsi/RhCMV
singularity shell --no-home --bind ${wd}:${wd}:rw ${sif}
R

```

#### Paths & Libraries {-}

```{r message=FALSE, warning=FALSE}
startdt <- date() # start time

library(tidyverse)
library(DT)
library(RColorBrewer)
library(Biostrings)
library(data.table)
library(glue)
library(openxlsx)

wd <- "/hpc/group/chsi/RhCMV"
source(file.path(wd, "Code", "config.R"))
source(file.path(wd, "Code","functions","utilities.R"))
source(file.path(wd, "Code","functions","strainSNP_utilities.R"))

# output directory for intermediates generated by this script
procdir.6filtsnvs

# directory with inputs generated by 2_nucdiff.sh
procdir.2nucdiff
gff.cols <- c("seqid","source","type","start","end",
              "score","strand","phase","attributes")

# directory with inputs generated by 3_get-mismatches.*
procdir.3mismatches

# directory with inputs generated by 5_prep-fractionDFs.Rmd
procdir.5fractiondfs

# Table and Figure numbers
# for newlab(), see functions/utilities.R
figlab <- "Figure 0"
tablab <- "Table 0"

```

#### Functions {-}

```{r}
# Processing/Refine strain SNP positions
summarize_windows <- function(dna, snp.positions, windf){
  
  # get the total number of positions in the reference strain
  npositions <- length(dna[[1]])
  
  # summarize local differences in windows that intersect each strain SNP
  df.l <- list()
  for(i in 1:length(snp.positions)){
    curr.pos <- snp.positions[i]
    df.l[[i]]<- windf %>%
      filter(start <= curr.pos & end >= curr.pos) %>%
      summarize(n = n(),
                mean.ndiffs = mean(nsnps),
                sd.ndiffs = sd(nsnps)) %>%
      mutate(snp.pos = curr.pos)
  }
  df.snpdiffs <- df.l %>%
    purrr::map_dfr(~data.frame(.x))
  
  return(df.snpdiffs)
  
}

lsf.str()

```

# Load data

## Load base-level strain frequency dataframes

- Base-level strain frequency dataframes generated in `5_prep-fractionDFs.Rmd`

```{r}
# VID
l.vid <- readRDSobj(fname = "l-vid.RDS", fdir = procdir.5fractiondfs)

# UCD
l.ucd <- readRDSobj(fname = "l-ucd.RDS", fdir = procdir.5fractiondfs)

tablab <- newlab(tablab, caption = "First 10 rows of 'plotdf' for 1 sample aligned to UCD")
head(l.ucd$`KH14P14`$plotdf) %>%
  simple.DT(title = tablab) %>%
  DT::formatRound(digits = 2, columns = "log2coverage")

```

## Load nucdiff local differences and identify strain SNPs

- Local differences between the reference strains were identified using Nucdiff, see `2_nucdiff.sh`
- Local differences that are single nucleotide polymorphisms (SNPs) are selected using the custom function `get_strainSNPs` defined in `functions/strainSNP_utilities.R`

```{r}
## VID ##---------------------------------------------------##
sSNPbases.vid <- load_select_SNPbases(aln.type = "R_VID_Q_UCD", procdir.2nucdiff, gff.cols)
nrow(sSNPbases.vid) # 2862

## UCD ##---------------------------------------------------##
sSNPbases.ucd <- load_select_SNPbases(aln.type = "Q_VID_R_UCD", procdir.2nucdiff, gff.cols)
nrow(sSNPbases.ucd) # 2865

tablab <- newlab(tablab, caption = "First 6 rows of strain SNPs in UCD")
head(sSNPbases.ucd) %>%
  knitr::kable(caption = tablab)

```

## Load sliding-window counts

- Sliding-window counts for each reference strain were generated in `3_get-mismatches.R/slurm`

```{r}
# load sliding-window counts from 3_get-mismatches.R/slurm

## VID ##---------------------------------------------------##
aln.type <- "R_VID_Q_UCD"
windf.vid <- readRDS(file = file.path(procdir.3mismatches, 
                                  paste0("windf_", aln.type,".RDS")))

## UCD ##---------------------------------------------------##
aln.type <- "Q_VID_R_UCD"
windf.ucd <- readRDS(file = file.path(procdir.3mismatches, 
                                  paste0("windf_", aln.type,".RDS")))

```

## Load reference genome fastas

- Reference genome fastas were downloaded from NCBI (`data-transfer/downlaod-NCBI-assemblies.sh`) and are used to check the total number of genome positions in the `summarize_windows` function

```{r}
# load the reference fasta

## VID ##---------------------------------------------------##
vidref.fna
dna.vid <- readDNAStringSet(vidref.fna)

## UCD ##---------------------------------------------------##
ucdref.fna
dna.ucd <- readDNAStringSet(ucdref.fna)

```


# Refine strain SNP positions {.tabset}

Refine the **set of strain SNP positions across all samples** based on the following criteria:

1. *Density of local differences*: Position has **fewer than 5** average local differences between reference strains in 100 bp sliding windows

2. *Alternate strain inoculum coverage*: Position has coverage **greater than 250** reads in the sequenced inoculum sample when mapped to the alternate reference strain, i.e. ...
    + UCD52 sequenced inoculum mapped to VID965 reference genome 
    + VID965 sequenced inoculum mapped to UCD52 reference genome  

3. *Alternate strain inoculum frequency*: Position has **less than 1% alternate strain frequency** in the sequenced inoculum samples. Positions with high alternate strain frequencies in the sequenced inoculum indicate heterogeneity in the inoculum which could bias strain frequency estimates in biological samples

## Density of local differences

- The number of local differences between strains at each genome position in 100bp sliding windows has been loaded from `3_get-mismatches.R/slurm`. 
- For each strain SNP position, identify all the 100bp windows that intersect that position. Summarize the mean number of local differences in all windows that intersect the SNP position.
- Save the results to `df-snpdiffs-<REf>.RDS`

```{r}
preproc <- FALSE

if (preproc) {
  
  ## VID ##---------------------------------------------------##
  df.snpdiffs.vid <- summarize_windows(dna = dna.vid, 
                                       snp.positions = sSNPbases.vid$POS, 
                                       windf = windf.vid)
  length(sSNPbases.vid$POS) # 2862
  nrow(df.snpdiffs.vid) # 2862
  
  ## UCD ##---------------------------------------------------##
  df.snpdiffs.ucd <- summarize_windows(dna = dna.ucd, 
                                       snp.positions = sSNPbases.ucd$POS, 
                                       windf = windf.ucd)
  length(sSNPbases.ucd$POS) # 2865
  nrow(df.snpdiffs.ucd) # 2865
  
  # save objects
  saveRDSobj(df.snpdiffs.vid, outdir = procdir.6filtsnvs)
  saveRDSobj(df.snpdiffs.ucd, outdir = procdir.6filtsnvs)
  
}else{
  
  df.snpdiffs.vid <- readRDSobj(fname = "df-snpdiffs-vid.RDS", fdir = procdir.6filtsnvs)
  df.snpdiffs.ucd <- readRDSobj(fname = "df-snpdiffs-ucd.RDS", fdir = procdir.6filtsnvs)
  
}

tablab <- newlab(tablab, caption = "First 6 rows of 'df.snpdiffs.ucd' for UCD")
head(df.snpdiffs.ucd) %>%
  knitr::kable(caption = tablab)

```

- Plot the number of local differences in 100bp windows across the genome

```{r}
plotit <- F
if(plotit){
  
  ## VID ##---------------------------------------------------##
figlab <- newlab(figlab, caption = "Local differences in 100bp sliding windows\n across genome (VID965)")
windf.vid %>%
  ggplot(aes(x = start, y = nsnps)) +
  geom_bar(stat = "identity") +
      theme_bw() +
      theme(axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
      xlab("Genome position (window start)") +
      ylab("Local strain differences in 100bp window") +
  ggtitle(figlab)

## UCD ##---------------------------------------------------##
figlab <- newlab(figlab, caption = "Local differences in 100bp sliding windows\n across genome (UCD52)")
windf.ucd %>%
  ggplot(aes(x = start, y = nsnps)) +
  geom_bar(stat = "identity") +
      theme_bw() +
      theme(axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
      xlab("UCD52 genome position") +
      ylab("Local strain differences in 100bp window") +
  ggtitle(figlab)
  
}

```

- Plot the distribution of mean local differences around each strain SNP

```{r}

## VID ##---------------------------------------------------##
figlab <- newlab(figlab, caption = "Histogram of mean local differences\n around strain SNPs (VID965)")
df.snpdiffs.vid %>%
  ggplot(aes(x = mean.ndiffs)) +
  geom_histogram() +
  ggtitle(figlab) +
  geom_vline(xintercept = 5, color = "blue")

## UCD ##---------------------------------------------------##
figlab <- newlab(figlab, caption = "Histogram of mean local differences\n around strain SNPs (UCD52)")
df.snpdiffs.ucd %>%
  ggplot(aes(x = mean.ndiffs)) +
  geom_histogram() +
  ggtitle(figlab)+
  geom_vline(xintercept = 5, color = "blue")

```

- Retain strain SNP positions with **fewer than 5** mean local differences in the 100bp windows that intersect each SNP

```{r}
# Retain strain SNP positions with **fewer than 5** mean local differences in the 100bp windows that intersect each SNP
max.localdiffs <- 5

## VID ##---------------------------------------------------##
df.snpdiffs.vid1 <- df.snpdiffs.vid %>%
  mutate(pass.localdiffs = mean.ndiffs < max.localdiffs)

tablab <- newlab(tablab, caption = "Mapped to VID965: Number of strain SNPs that pass the local differences density filter")
df.snpdiffs.vid1 %>%
  group_by(pass.localdiffs) %>%
  summarize(n.snps = n()) %>%
  simple.DT(title = tablab)

## UCD ##---------------------------------------------------##
df.snpdiffs.ucd1 <- df.snpdiffs.ucd %>%
  mutate(pass.localdiffs = mean.ndiffs < max.localdiffs)

tablab <- newlab(tablab, caption = "Mapped to UCD52: Number of strain SNPs that pass the local differences density filter")
df.snpdiffs.ucd1 %>%
  group_by(pass.localdiffs) %>%
  summarize(n.snps = n()) %>%
  simple.DT(title = tablab)

```

## Alternate strain inoculum coverage

- Identify alternate strain inoculum coverage at strain SNP positions

```{r}
# Identify alternate strain inoculum coverage at strain SNP positions
inoc.vid <- l.vid[["UCD52mda"]]$plotdf %>%
  dplyr::select(pos, coverage)
inoc.ucd <- l.ucd[["965BACmda"]]$plotdf %>%
  dplyr::select(pos, coverage)

tablab <- newlab(tablab, caption = "First 6 rows of VID inoculum coverage mapped to UCD52")
inoc.ucd %>%
  head() %>%
  knitr::kable(caption = tablab)

```

- Plot the distribution of alternate strain inoculum coverage at strain SNPs

```{r}

## VID ##---------------------------------------------------##
figlab <- newlab(figlab, caption = "Histogram of UCD inoculumn coverage \nmapped to VID at strain SNPs")
inoc.vid %>%
  ggplot(aes(x = coverage)) +
  geom_histogram() +
  ggtitle(figlab) +
  geom_vline(xintercept = 250, color = "blue")

figlab <- newlab(figlab, caption = "Filter to coverage less than 1000: Histogram of UCD inoculumn coverage \nmapped to VID at strain SNPs")
inoc.vid %>%
  filter(coverage < 1000) %>%
  ggplot(aes(x = coverage)) +
  geom_histogram() +
  ggtitle(figlab) +
  geom_vline(xintercept = 250, color = "blue") 


## UCD ##---------------------------------------------------##
figlab <- newlab(figlab, caption = "Histogram of VID inoculumn coverage \nmapped to UCD at strain SNPs")
inoc.ucd %>%
  ggplot(aes(x = coverage)) +
  geom_histogram() +
  ggtitle(figlab)+
  geom_vline(xintercept = 250, color = "blue")

figlab <- newlab(figlab, caption = "Filter to coverage less than 1000: Histogram of VID inoculumn coverage \nmapped to UCD at strain SNPs")
inoc.ucd %>%
  filter(coverage < 1000) %>%
  ggplot(aes(x = coverage)) +
  geom_histogram() +
  ggtitle(figlab)+
  geom_vline(xintercept = 250, color = "blue")


```


- Retain strain SNP positions with **greater than 250** reads

```{r}
# Retain strain SNP positions with **greater than 250** reads
min.inocov <- 250

## VID ##---------------------------------------------------##
df.snpdiffs.vid2 <- df.snpdiffs.vid1 %>%
  left_join(inoc.vid, by = c("snp.pos" = "pos")) %>%
  mutate(pass.inocov = coverage > min.inocov)

tablab <- newlab(tablab, caption = "Mapped to VID965: Number of strain SNPs that pass the alt strain inoculum coverage filter")
df.snpdiffs.vid2 %>%
  group_by(pass.inocov) %>%
  summarize(n.snps = n()) %>%
  simple.DT(title = tablab)

## UCD ##---------------------------------------------------##
df.snpdiffs.ucd2 <- df.snpdiffs.ucd1 %>%
  left_join(inoc.ucd, by = c("snp.pos" = "pos")) %>%
  mutate(pass.inocov = coverage > min.inocov)

tablab <- newlab(tablab, caption = "Mapped to UCD52: Number of strain SNPs that pass the alt strain inoculum coverage filter")
df.snpdiffs.ucd2 %>%
  group_by(pass.inocov) %>%
  summarize(n.snps = n()) %>%
  simple.DT(title = tablab)

```

## Alternate strain inoculum frequency

- Get alternate strain base-level frequencies for the sequenced inoculum samples, i.e. ...

  - UCD frequency in sequenced "965BACmda" inoculum mapped to VID; pull from `alt.base.frac`
  - VID frequency in sequenced "UCD52mda" inoculum mapped to VID; pull from `ref.base.frac`
  - UCD frequency in sequenced "965BACmda" inoculum mapped to UCD; pull from `ref.base.frac`
  - VID frequency in sequenced "UCD52mda" inoculum mapped to UCD; pull from `alt.base.frac`
  
```{r}
# Get alternate strain base-level frequencies for the sequenced inoculum samples
## VID ##---------------------------------------------------##
curr.sample <- "965BACmda"
inocF.vid1 <- l.vid[[curr.sample]]$plotdf %>%
  dplyr::select(pos, 
                "frac"= alt.base.frac) %>%
  mutate(reference = "VID965",
         inoculum.type = curr.sample)

curr.sample <- "UCD52mda"
inocF.vid2 <- l.vid[[curr.sample]]$plotdf %>%
  dplyr::select(pos, 
                "frac"= ref.base.frac) %>%
  mutate(reference = "VID965",
         inoculum.type = curr.sample)

inocF.vid <- inocF.vid1 %>%
  bind_rows(inocF.vid2) %>%
  pivot_wider(id_cols = c(pos, reference), 
              names_from = inoculum.type,
              values_from = frac)

## UCD ##---------------------------------------------------##
curr.sample <- "965BACmda"
inocF.ucd1 <- l.ucd[[curr.sample]]$plotdf %>%
  dplyr::select(pos, "frac"= ref.base.frac) %>%
  mutate(reference = "UCD52",
         inoculum.type = curr.sample, # VID965
         var.type.name = "UCD52") # not VID965

curr.sample <- "UCD52mda"
inocF.ucd2 <- l.ucd[[curr.sample]]$plotdf %>%
  dplyr::select(pos, "frac"= alt.base.frac) %>%
  mutate(reference = "UCD52",
         inoculum.type = curr.sample, # UCD52
         var.type.name = "VID965") # not UCD52

inocF.ucd <- inocF.ucd1 %>%
  bind_rows(inocF.ucd2) %>%
  pivot_wider(id_cols = c(pos, reference), 
              names_from = inoculum.type,
              values_from = frac)


tablab <- newlab(tablab, caption = "First 6 rows of UCD frequency in sequenced VID inoculum mapped to UCD")
inocF.ucd1 %>%
  head() %>%
  knitr::kable(caption = tablab)

tablab <- newlab(tablab, caption = "First 6 rows of VID frequency in sequenced UCD inoculum mapped to UCD")
inocF.ucd2 %>%
  head() %>%
  knitr::kable(caption = tablab)

```

- Plot the distribution of alternate strain inoculum frequencies at strain SNPs

```{r}

## VID ##---------------------------------------------------##
figlab <- newlab(figlab, caption = "Histogram of UCD frequency in VID inoculum mapped to VID")
inocF.vid1 %>%
  filter(!is.na(frac)) %>%
  ggplot(aes(x = frac)) +
  geom_histogram() +
  ggtitle(figlab) +
  geom_vline(xintercept = 0.01, color = "blue")

figlab <- newlab(figlab, caption = "Histogram of VID frequency in UCD inoculum mapped to VID")
inocF.vid2 %>%
  filter(!is.na(frac)) %>%
  ggplot(aes(x = frac)) +
  geom_histogram() +
  ggtitle(figlab) +
  geom_vline(xintercept = 0.01, color = "blue")


## UCD ##---------------------------------------------------##
figlab <- newlab(figlab, caption = "Histogram of UCD frequency in VID inoculum mapped to UCD")
inocF.ucd1 %>%
  filter(!is.na(frac)) %>%
  ggplot(aes(x = frac)) +
  geom_histogram() +
  ggtitle(figlab) +
  geom_vline(xintercept = 0.01, color = "blue")

figlab <- newlab(figlab, caption = "Histogram of VID frequency in UCD inoculum mapped to UCD")
inocF.ucd2 %>%
  filter(!is.na(frac)) %>%
  ggplot(aes(x = frac)) +
  geom_histogram() +
  ggtitle(figlab) +
  geom_vline(xintercept = 0.01, color = "blue")


```


- Retain strain SNP positions with less than 1% alternate strain frequency in the sequenced inoculum samples. Positions with high alternate strain frequencies in the sequenced inoculum indicate heterogeneity in the inoculum. For more investigation into these positions, see `gviz.html`

```{r}
# Retain strain SNP positions with less than 1% alternate strain frequency in the sequenced inoculum samples
max.inoAltfreq <- 0.01 # 1%

## VID ##---------------------------------------------------##
# If the alt freq > 1% in sequenced UCD or VID inoculum, we don't want the position
df.snpdiffs.vid3 <- df.snpdiffs.vid2 %>%
  left_join(inocF.vid, by = c("snp.pos" = "pos")) %>%
  mutate(
    fail.inoVID = ifelse(!is.na(`965BACmda`) & 
                           `965BACmda` >= max.inoAltfreq, 
                         TRUE,
                         FALSE),
    fail.inoUCD = ifelse(!is.na(`UCD52mda`) & 
                           `UCD52mda` >= max.inoAltfreq, 
                         TRUE,
                         FALSE),
    fail.inoAltfreq = fail.inoVID | fail.inoUCD,
    pass.inoAltfreq = !fail.inoAltfreq)

tablab <- newlab(tablab, caption = "Mapped to VID965: Number of strain SNPs that pass the alt strain inoculum frequency filter")
df.snpdiffs.vid3 %>%
  group_by(fail.inoVID, fail.inoUCD, 
           fail.inoAltfreq, pass.inoAltfreq) %>%
  summarize(n.snps = n()) %>%
  simple.DT(title = tablab)

## UCD ##---------------------------------------------------##
df.snpdiffs.ucd3 <- df.snpdiffs.ucd2 %>%
  left_join(inocF.ucd, by = c("snp.pos" = "pos")) %>%
  mutate(
    fail.inoVID = ifelse(!is.na(`965BACmda`) & 
                           `965BACmda` >= max.inoAltfreq, 
                         TRUE,
                         FALSE),
    fail.inoUCD = ifelse(!is.na(`UCD52mda`) & 
                           `UCD52mda` >= max.inoAltfreq, 
                         TRUE,
                         FALSE),
    fail.inoAltfreq = fail.inoVID | fail.inoUCD,
    pass.inoAltfreq = !fail.inoAltfreq)

tablab <- newlab(tablab, caption = "Mapped to UCD52: Number of strain SNPs that pass the alt strain inoculum frequency filter")
df.snpdiffs.ucd3 %>%
  group_by(fail.inoVID, fail.inoUCD, fail.inoAltfreq, pass.inoAltfreq) %>%
  summarize(n.snps = n()) %>%
  simple.DT(title = tablab)

```

# Summarize filtered strain SNPs

- Count the number of strain SNPs filtered out by combinations of the three criteria. Save the inventory of filtering SNPs for plotting.

```{r}

## VID ##---------------------------------------------------##
nrow(df.snpdiffs.vid3) # 2862

tablab <- newlab(tablab, caption = "Mapped to VID965: Summary of filtered strain SNPs")
df.snpdiffs.vid3 %>%
  group_by(pass.localdiffs, 
                pass.inocov,
                pass.inoAltfreq) %>%
  summarize(n.snps = n()) %>%
  simple.DT(title = tablab)

viddfout <- file.path(procdir.6filtsnvs, "df-snpdiffs-vid3.RDS")
saveRDS(df.snpdiffs.vid3, viddfout)

## UCD ##---------------------------------------------------##
nrow(df.snpdiffs.ucd3) # 2865

tablab <- newlab(tablab, caption = "Mapped to UCD52: Summary of filtered strain SNPs")
df.snpdiffs.ucd3 %>%
  group_by(pass.localdiffs, 
                pass.inocov,
                pass.inoAltfreq) %>%
  summarize(n.snps = n()) %>%
  simple.DT(title = tablab)

ucddfout <- file.path(procdir.6filtsnvs, "df-snpdiffs-ucd3.RDS")
saveRDS(df.snpdiffs.ucd3, ucddfout)




tablab <- newlab(tablab, caption = "First 6 rows of strain SNP inventory mapped to UCD")
df.snpdiffs.ucd3 %>%
  head() %>%
  simple.DT(title = tablab)


```

# Apply filters

```{r}
# --- pull strainSNP vectors for VID and UCD
snp.positions.filt.vid <- df.snpdiffs.vid3 %>%
  filter(pass.localdiffs, pass.inocov, pass.inoAltfreq) %>%
  pull(snp.pos)
length(snp.positions.filt.vid) # 862

snp.positions.filt.ucd <- df.snpdiffs.ucd3 %>%
  filter(pass.localdiffs, pass.inocov, pass.inoAltfreq) %>%
  pull(snp.pos)
length(snp.positions.filt.ucd) # 850


# --- filter positions within each sample

samples <- names(l.vid) # same for l.ucd
l.vid.filt <- list()
l.ucd.filt <- list()

for (i in 1:length(samples)) {
  
  name <- samples[i]
  l.vid.filt[[name]] <- l.vid[[name]] %>%
    map(~ filter(.x, pos %in% snp.positions.filt.vid))
  l.ucd.filt[[name]] <- l.ucd[[name]] %>%
    map(~ filter(.x, pos %in% snp.positions.filt.ucd))
  
  l.vid.filt[[name]]$plotdf <- l.vid.filt[[name]]$plotdf %>%
    mutate(lowcoverage = ifelse(coverage < 100, TRUE, FALSE))
  l.ucd.filt[[name]]$plotdf <- l.ucd.filt[[name]]$plotdf %>%
    mutate(lowcoverage = ifelse(coverage < 100, TRUE, FALSE))
  
}

saveRDSobj(l.vid.filt, outdir = procdir.6filtsnvs)
saveRDSobj(l.ucd.filt, outdir = procdir.6filtsnvs)

```

# Session information

```{r, results='hold'}
sessionInfo()

print(glue("Start Time: ",startdt))
print(glue("End Time: ",date()))

```
