---
title: "Results"
author: "DCI Bioinformatics"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    code_folding: hide
    number_sections: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE, 
  comment = '',
  results = "asis",
  warning = FALSE, 
  message = FALSE, 
  fig.align = 'center',
  size = 'footnotesize',
  fig.asp = 0.5, 
  out.width = "80%",
  fig.width = 7,
  tidy.opts = list(keep.blank.line = FALSE,
                   width.cutoff = 70)
)
```

```{r dtsetup, include=FALSE}
# init step to make sure that DT dependencies are loaded
# if this is not pre-set, tables may not display properly from
# within for loops

htmltools::tagList(
  DT::datatable(
    iris, 
    rownames = FALSE,
    filter = "top",
    extensions = 'Buttons', 
    options = list(
      dom = 'Bfrtip',
      scrollX = TRUE,
      buttons = list(
        list(extend = 'colvis', 
             columns = c(0,1,2,3,4)),
        list(extend = "colvisGroup",
             text = "Show All",
             show = ":hidden")),
      columnDefs = list(
        list(visible = FALSE,
             targets = c(0,1))
      )
    )
  )
)

```

# Notes

Reference strains:

- VID965-FL(GCA_027929765.1)
- UCD52 (GCA_027930255.1)

Alignment types:

- *RvidQucd* = UCD52 (query) was aligned to VID965 (reference)
- *QvidRucd* = VID965 (query) was aligned to UCD52 (reference)

# Prerequisites

#### Container {-}

```{bash eval=F}

sif=/opt/apps/containers/community/ACC/dcibioinformaticsR-v2.3.sif
wd=/hpc/group/chsi/RhCMV
singularity shell --no-home --bind ${wd}:${wd}:rw ${sif}
R

```

#### Paths & Libraries {-}

```{r message=FALSE, warning=FALSE}
startdt <- date() # start time

library(tidyverse)
library(DT)
library(RColorBrewer)
library(Biostrings)
library(data.table)
library(glue)
library(openxlsx)

# update patchwork version to use "free" functions
#install.packages('patchwork')
library(patchwork)
# packageVersion('patchwork') # 1.3.0

library(showtext)
# font_files() # no Arial?
font_add_google("Arimo") # this is similar
# NOTES:
showtext_auto() # turn fonts on
#showtext_auto(enable = FALSE) # turn fonts off
showtext_opts(dpi = 96) # for printing to RStudio screen

wd <- "/hpc/group/chsi/RhCMV"
source(file.path(wd, "Code", "config.R"))
source(file.path(wd, "Code/functions", "utilities.R"))
source(file.path(wd, "Code","functions","strainSNP_utilities.R"))

# output directory for intermediates generated by this script
procdir.7results

# output directory for resulting figures and tables generated by this script
resultdir

# directory with inputs generated by 2_nucdiff.sh
procdir.2nucdiff
gff.cols <- c("seqid","source","type","start","end",
              "score","strand","phase","attributes")

# directory with inputs generated by 6_filtering-SNVs.Rmd
procdir.6filtsnvs

# If you would like to re-print tables/figures, update to T
printtables <- F
printpngs <- F
printpdfs <- F

```

#### Formatting {-}

```{r message=FALSE, warning=FALSE}
# Table and Figure numbers
# for newlab(), see functions/utilities.R
figlab <- "Figure 0"
tablab <- "Table 0"

sel.constant <- 0.5 # constant to add in the Log2(coverage + constant) calculation

# Reference strain colors
strain.cols <- c("RhCMV UCD52" = "blue",
                 "FL-RhCMV" = "orange") # used in sample-level plots

strain.cols2 <- c("RhCMV UCD52" = "blue",
                  "FL-RhCMV" = "orange",
                  "Other" = "red",
                  "Low Coverage" = "gray") # used in base-level plots and pre-filt panels

strain.cols3 <- c("RhCMV UCD52" = "blue",
                  "FL-RhCMV" = "orange",
                  "Other" = "red",
                  "Filtered Position" = "gray") # used in post-filt panels

strain.cols.filt <- c("RhCMV UCD52" = "blue",
                      "FL-RhCMV" = "orange",
                      "Other" = "red",
                      "Zero Coverage" = "gray") # used in pre-filt panels

strain.cols.Xgenome <- c("RhCMV UCD52" = "blue",
                         "FL-RhCMV" = "orange",
                         "Other" = "red") # used in whole-genome extrapolation

# Pass filter colors
pass.cols <- c("FALSE" = "gray",
               "TRUE" = "black") # used in filtering panels

# Levels
levels.jitter <- c("P4 (M)", "P7 (M)", "P14 (M)", "AF77 (F)",
                   "P82 (M)", "AF82 (F)", "heart (F)", "kidney (F)",
                   "lung (F)", "basalganglion (F)", "parietalcortex (F)")
levels.mixing <- c("1FL-10UCD", "1FL-5UCD", "1FL-1UCD", "5FL-1UCD", "10FL-1UCD")


# Plot printing
set.dpi <- 1200

# Sample-level estimate plots
ht.samplelv <- 3
wt.samplelv <- 5

# Base-level estimate plots
ht.baselv <- 7
wt.baselv <- 7

# Barcode plot
ht.barcode <- 1
wt.barcode <- 6

# Filtering scheme panels
wt.filtpanel <- 4
ht.filtpanel <- 2
wt.legend <- 2
ht.legend <- 2

```

Function to print plots:
```{r}
print_baselevel_plot <- function(p, outfile.base, printpngs, printpdfs, set.dpi, height, width){
  
  # outfile.base <- file.path(resultdir, "sample-level-estimate-plots","strain-freq-ucd-nucdiff-mixing")
  
  if (printpngs) {
    
    showtext_opts(dpi = set.dpi)
    outfile <- glue("{outfile.base}.png")
    ggsave(filename = outfile, plot = p,
           height = height,
           width = width,
           dpi = set.dpi)
    showtext_opts(dpi = 96) # reset
    
    print(outfile)
  }
  if (printpdfs) {
    
    outfile <- glue("{outfile.base}.pdf")
    ggsave(filename = outfile, plot = p,
           height = height,
           width = width)
    print(outfile)
    
  }
  
}

```

# Load data and prep

## Load unfiltered fractionDFs

- FractionDFs were generated in `5_prep-fractionDFs.Rmd`
- These are used in the "Make filter scheme panels" and "Base-level estimate plots -- all genome positions" sections below

```{r}
l.vid <- readRDSobj(fname = "l-vid.RDS", fdir = procdir.5fractiondfs)
l.ucd <- readRDSobj(fname = "l-ucd.RDS", fdir = procdir.5fractiondfs)

# --- update strain labels in unfiltered dfs
all.samples <- names(l.vid)
for (i in 1:length(all.samples)) {
  # --- UCD
  name <- all.samples[i]
  l.ucd[[name]]$plotdf.fracs <- l.ucd[[name]]$plotdf.fracs %>%
    mutate(var.type.2 = case_when(
      var.type.name == "UCD52" ~ "RhCMV UCD52",
      var.type.name == "VID965" ~ "FL-RhCMV",
      var.type.name == "other.frac" ~ "Other"
    ))
  
  # --- VID
  l.vid[[name]]$plotdf.fracs <- l.vid[[name]]$plotdf.fracs %>%
    mutate(var.type.2 = case_when(
      var.type.name == "UCD52" ~ "RhCMV UCD52",
      var.type.name == "VID965" ~ "FL-RhCMV",
      var.type.name == "other.frac" ~ "Other"
    ))
}

```

## Load filtered SNP inventories

- Filtered SNP inventories were generated in `6_filtering-SNVs.Rmd`
- These are used in the section below "Make filter scheme panels"

```{r}
df.snpdiffs.vid3 <- readRDSobj(fname = "df-snpdiffs-vid3.RDS", fdir = procdir.6filtsnvs)
df.snpdiffs.ucd3 <- readRDSobj(fname = "df-snpdiffs-ucd3.RDS", fdir = procdir.6filtsnvs)

```

## Load filtered fractionDFs

- Filtered fractionDFs were generated in `6_filtering-SNVs.Rmd`
- These are used in the sections below "Sample-level summary tables", "Sample-level estimate plots", "Base-level estimate plots", and "Base-level estimate plots- all geneome positions"

```{r}
l.vid.filt <- readRDSobj(fname = "l-vid-filt.RDS", fdir = procdir.6filtsnvs)
l.ucd.filt <- readRDSobj(fname = "l-ucd-filt.RDS", fdir = procdir.6filtsnvs)

mixing.sams <- all.samples[grepl("^[1,5]", all.samples)] 
kh14.sams <- all.samples[!all.samples %in% mixing.sams]

# --- update strain labels in filtered dfs
for (i in 1:length(all.samples)) {
  
  name <- all.samples[i]
  # --- UCD
  lowcov.ucd <- l.ucd.filt[[name]]$plotdf %>%
    filter(lowcoverage) %>%
    pull(pos)
  l.ucd.filt[[name]]$plotdf.fracs <- l.ucd.filt[[name]]$plotdf.fracs %>%
    mutate(var.type.2 = case_when(
      var.type.name == "UCD52" ~ "RhCMV UCD52",
      var.type.name == "VID965" ~ "FL-RhCMV",
      var.type.name == "other.frac" ~ "Other"
    ),
    var.type.2 = ifelse(pos %in% lowcov.ucd, "Low Coverage", var.type.2))
  
  # --- VID
  lowcov.vid <- l.vid.filt[[name]]$plotdf %>%
    filter(lowcoverage) %>%
    pull(pos)
  l.vid.filt[[name]]$plotdf.fracs <- l.vid.filt[[name]]$plotdf.fracs %>%
    mutate(var.type.2 = case_when(
      var.type.name == "UCD52" ~ "RhCMV UCD52",
      var.type.name == "VID965" ~ "FL-RhCMV",
      var.type.name == "other.frac" ~ "Other"
    ),
    var.type.2 = ifelse(pos %in% lowcov.vid, "Low Coverage", var.type.2))
}

# --- subset filt objs for sections below
l.vid.kh14 <- l.vid.filt[kh14.sams]
l.ucd.kh14 <- l.ucd.filt[kh14.sams]
l.vid.mixing <- l.vid.filt[mixing.sams]
l.ucd.mixing <- l.ucd.filt[mixing.sams]

```

## Load strain SNPs with VID as the reference

- Local differences between the reference strains were identified using Nucdiff, see `2_nucdiff.sh`
- Local differences that are single nucleotide polymorphisms (SNPs) are selected using the custom function `get_strainSNPs` defined in `functions/strainSNP_utilities.R`
- This is used in the section below "Plot nucdiff SNPs along genome (R_VID_Q_UCD)"

```{r}
## VID ##---------------------------------------------------##
sSNPbases.vid <- load_select_SNPbases(aln.type = "R_VID_Q_UCD", procdir.2nucdiff, gff.cols)
nrow(sSNPbases.vid) # 2862

```

# Sample-level summary tables

- Compile the summarize table for all samples

```{r}
summtab.l <- list()
#i<-1
for (i in 1:length(all.samples)) {
  
  name <- all.samples[i]
  
  ## VID ##---------------------------------------------------##
  snpvec.vid <- l.vid.filt[[name]]$plotdf %>%
    filter(lowcoverage == FALSE) %>%
    pull(pos)
  summ.vid <- l.vid.filt[[name]]$plotdf.fracs %>%
    filter(pos %in% snpvec.vid) %>%
    group_by(var.type.name) %>%
    summarize(n = n(),
              mean = mean(frac),
              se = sd(frac)/sqrt(n)) %>% 
    mutate(alignment.type.type = "RvidQucd")
  
  ## UCD ##---------------------------------------------------##
  snpvec.ucd <- l.ucd.filt[[name]]$plotdf %>%
    filter(lowcoverage == FALSE) %>%
    pull(pos)
  summ.ucd <- l.ucd.filt[[name]]$plotdf.fracs %>%
    filter(pos %in% snpvec.ucd) %>%
    group_by(var.type.name) %>%
    summarize(n = n(),
              mean = mean(frac),
              se = sd(frac)/sqrt(n)) %>%
    mutate(alignment.type.type = "QvidRucd")
  
  summtab <- summ.vid %>%
    bind_rows(summ.ucd) %>%
    arrange(var.type.name) 
  
  summtab.l[[i]] <- summtab
  
}

names(summtab.l) <- all.samples
supptab <- summtab.l %>%
  purrr::map_dfr(~data.frame(.x), .id = "sample") 

```

- Print tidy versions 

```{r}
supptab.tidy <- supptab %>%
  mutate(mean.perc = signif(mean * 100, digits = 4),
         se.perc = signif(se * 100, digits = 4),
         mean.text = paste0(mean.perc, " (", se.perc, ")" )) %>%
  mutate(alignment.type.type = ifelse(alignment.type.type == "RvidQucd", 
                                      "FL-RhCMV",
                                      ifelse(alignment.type.type == "QvidRucd", 
                                             "UCD52", 
                                             NA))) %>%
  dplyr::select(sample, alignment.type.type, n, 
                var.type.name, mean.text) %>%
  pivot_wider(id_cols = c(sample, alignment.type.type, n),
              names_from = var.type.name,
              values_from = mean.text) %>%
  dplyr::select("Sample" = sample, 
                "Mapping reference" = alignment.type.type,
                 "SNPs used for analysis" = n,
                "UCD52 mean strain frequency (SE)" = UCD52,
                "FL mean strain frequency (SE)" = VID965,
                "Other mean strain frequency (SE)" = other.frac)

tablab <- newlab(tablab, caption = "Sample-level summary table, all samples")
supptab.tidy %>%
  simple.DT(title = tablab)

if (printtables) {
  
  # print only samples included in manuscript
  supptab.tidy %>%
    filter(!Sample %in% mixing.sams) %>%
    write.csv(file.path(resultdir, "mean-strain-freq-table_manuSamples.csv"))
  
  # print only mixing samples
  supptab.tidy %>%
    filter(Sample %in% mixing.sams) %>%
    write.csv(file.path(resultdir, "mean-strain-freq-table_mixingSamples.csv"))
}

```

- Range of SNPs used in analysis across samples presented in the manuscript

```{r}
supptab.tidy %>%
  filter(!Sample %in% mixing.sams) %>%
  pull(`SNPs used for analysis`) %>%
  range() # 14, 862

```

# Sample-level estimate plots

Functions:

```{r}
make_strainplots_jitter <- function(l, 
                                    exclude = c("965BACmda", "UCD52mda", "P77"), 
                                    figlab = NULL,
                                    plot.levels = NULL,
                                    strain.cols){
  
  # l <- l.vid.kh14
  # l <- l.vid.mixing
  # any(grepl("KH14", names(l)))
  
  if (any(grepl("KH14", names(l)))) {
  
    plotdf <- data.frame()
    for (i in 1:length(l)) {
      
      lst <- l[[i]]
      name <- names(l)[i]
      
      snpvec <- lst$plotdf %>%
        filter(lowcoverage == FALSE) %>%
        pull(pos) %>%
        unique()
      
      df <- lst$plotdf.fracs %>%
        transmute(sample = gsub("KH14", "", name),
                  plot.name = ifelse(grepl("^P", sample),
                                     paste0(sample, " (M)"),
                                     paste0(sample, " (F)")),
                  plot.name = factor(plot.name, levels = plot.levels),
                  pos,
                  freq = frac * 100,
                  strain = var.type.2) %>%
        filter(pos %in% snpvec,
               strain != "Other")
      
      plotdf <- rbind(plotdf, df)
      
    }
  
  } else {
    
    plotdf <- data.frame()
    for (i in 1:length(l)) {
      
      lst <- l[[i]]
      name <- names(l)[i]
      
      snpvec <- lst$plotdf %>%
        filter(lowcoverage == FALSE) %>%
        pull(pos) %>%
        unique()
      
      df <- lst$plotdf.fracs %>%
        transmute(sample = name,
                  plot.name = factor(sample, levels = plot.levels),
                  pos,
                  freq = frac * 100,
                  strain = var.type.2) %>%
        filter(pos %in% snpvec,
               strain != "Other")
      
      plotdf <- rbind(plotdf, df)
      
    }
    
  }
    
  plotdf <- plotdf %>%
    filter(!sample %in% exclude)
  
  mean.df <- plotdf %>%
    group_by(plot.name, strain) %>%
    summarize(mean.freq = mean(freq),
              se.freq = sd(freq) / sqrt(length(freq)),
              n = n())
  
  bar.df <- mean.df %>%
    select(plot.name) %>%
    distinct() %>%
    mutate(bar.height = 100)
  
  # --- mean + se plot
  dodge <- 0.25
  
  p <- mean.df %>%
    ggplot(aes(x = plot.name, 
               y = mean.freq, 
               color = strain)) +
    geom_bar(data = bar.df,
             stat = "identity",
             aes(x = plot.name,
                 y = bar.height),
             fill = "#F8F8F8",
             color = "black") +
    geom_point(data = plotdf,
               aes(x = plot.name,
                   y = freq,
                   fill = strain),
               inherit.aes = FALSE,
               shape = 16,
               color = "black",
               size = 0.5,
               alpha = 0.1,
               position = position_jitterdodge(jitter.width = 0.3,
                                               dodge.width = dodge * 3)) +
    geom_point(size = 2,
               pch = 1,
               position = position_dodge(width = dodge * 3)) +
    geom_errorbar(aes(ymin = mean.freq - se.freq,
                      ymax = mean.freq + se.freq),
                  position = position_dodge(width = dodge * 3)) +
    scale_color_manual(name = "Strain", 
                       values = strain.cols) +
    guides(fill = "none") +
    xlab("") +
    ylab("Inferred relative\nstrain frequency (%)") +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45,
                                     hjust = 1,
                                     vjust = 1),
          text = element_text(family = "Arimo")) +
    ggtitle(figlab)
  
}

```

## KH14 {.tabset}

### Aligned to VID965

```{r, fig.width=10, fig.asp=0.5}
figlab <- newlab(figlab, caption = "Sample-level estimates, aligned to VID965")
p <- make_strainplots_jitter(l = l.vid.kh14, 
                             figlab = NULL,
                             plot.levels = levels.jitter,
                             strain.cols = strain.cols)

p + ggtitle(figlab)

# Printing
outfile.base <- file.path(resultdir, "sample-level-estimate-plots","strain-freq-vid-nucdiff-n14")
print_baselevel_plot(p,
                     outfile.base, printpngs, printpdfs, set.dpi, 
                     height = ht.samplelv, 
                     width = wt.samplelv)
  
```

### Aligned to UCD52

```{r, fig.width=10, fig.asp=0.5}
figlab <- newlab(figlab, caption = "Sample-level estimates, aligned to UCD52")
p <- make_strainplots_jitter(l = l.ucd.kh14, 
                             figlab = NULL,
                             plot.levels = levels.jitter,
                             strain.cols = strain.cols)
p + ggtitle(figlab)


# Printing
outfile.base <- file.path(resultdir, "sample-level-estimate-plots","strain-freq-ucd-nucdiff-n14")
print_baselevel_plot(p,
                     outfile.base, printpngs, printpdfs, set.dpi, 
                     height = ht.samplelv, 
                     width = wt.samplelv)

```

## Mixing {.tabset}

### Aligned to VID965

```{r, fig.width=10, fig.asp=0.5}
figlab <- newlab(figlab, caption = "Sample-level estimates, aligned to VID965")
p <- make_strainplots_jitter(l = l.vid.mixing, 
                             figlab = NULL,
                             plot.levels = levels.mixing,
                             strain.cols = strain.cols)

p + ggtitle(figlab)

# Printing
outfile.base <- file.path(resultdir, "sample-level-estimate-plots","strain-freq-vid-nucdiff-mixing")
print_baselevel_plot(p,
                     outfile.base, printpngs, printpdfs, set.dpi, 
                     height = ht.samplelv, 
                     width = wt.samplelv)

```

### Aligned to UCD52

```{r, fig.width=10, fig.asp=0.5}
figlab <- newlab(figlab, caption = "Sample-level estimates, aligned to UCD52")
p <- make_strainplots_jitter(l = l.ucd.mixing, 
                             figlab = NULL,
                             plot.levels = levels.mixing,
                             strain.cols = strain.cols)

p + ggtitle(figlab)

# Printing
outfile.base <- file.path(resultdir, "sample-level-estimate-plots","strain-freq-ucd-nucdiff-mixing")
print_baselevel_plot(p,
                     outfile.base, printpngs, printpdfs, set.dpi, 
                     height = ht.samplelv, 
                     width = wt.samplelv)

```

# Base-level estimate plots

Functions:

```{r}
make_strainplots_panels <- function(lucd, lvid, sam, 
                                    colors = strain.cols2,
                                    constant = sel.constant){
  
  # test
  # sam <- "965BACmda"
  # lvid <- l.vid.filt
  # lucd <- l.ucd.filt
  
  lvid <- lvid[[sam]]
  lucd <- lucd[[sam]]
  if (sam %in% kh14.sams) {
    plot.name <- gsub("KH14", "", sam)
    if (grepl("^P", plot.name)) {
      plot.name <- paste0(plot.name, " (M)")
    } else if (!grepl("^P", plot.name) &
               !sam %in% c("965BACmda", "UCD52mda")) {
      plot.name <- paste0(plot.name, " (F)")
    }
  } else {
    plot.name <- sam
  }
  
  # --- VID panels
  p.frac.vid <- lvid$plotdf.fracs %>%
    arrange(pos) %>%
    mutate(frac.2 = ifelse(var.type.2 == "Low Coverage", 1/3, frac),
           pos.idx = rep(1:length(unique(pos)), each = 3),
           var.type.2 = factor(var.type.2, levels= names(colors))) %>%
    ggplot(aes(x = pos.idx, 
               y = frac.2, 
               fill = var.type.2, 
               color = var.type.2)) +
    geom_bar(stat = "identity", position = "stack") +
    theme_classic() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank()) +
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_continuous(expand = c(0, 0)) +
    scale_fill_manual(name = "",
                      values = colors) +
    scale_color_manual(name = "",
                       values = colors) +
    guides(color = "none") +
    xlab("SNV position") +
    ylab("Base fraction") +
    labs(title = plot.name,
         subtitle = "FL-RhCMV")
  
  p.cov.vid <- lvid$plotdf %>%
    mutate(pos.idx = 1:nrow(.)) %>%
    ggplot(aes(x = pos.idx, y = log2(coverage + constant))) +
    geom_bar(stat = "identity", color = "black") +
    theme_classic() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank()) +
    scale_x_continuous(expand = c(0, 0)) +
    ylim(c(-1, 17)) +
    xlab("SNV position") +
    ylab("Log2 coverage")
  
  p.frac.ucd <- lucd$plotdf.fracs %>%
    arrange(pos) %>%
    mutate(frac.2 = ifelse(var.type.2 == "Low Coverage", 1/3, frac),
           pos.idx = rep(1:length(unique(pos)), each = 3),
           var.type.2 = factor(var.type.2, levels= names(colors))) %>%
    ggplot(aes(x = pos.idx, y = frac.2, fill = var.type.2, color = var.type.2)) +
    geom_bar(stat = "identity", position = "stack") +
    theme_classic() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank()) +
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_continuous(expand = c(0, 0)) +
    scale_fill_manual(name = "",
                      values = colors) +
    scale_color_manual(name = "",
                       values = colors) +
    guides(color = "none") +
    xlab("SNV position") +
    ylab("Base fraction") +
    labs(subtitle = "RhCMV UCD52")
  
  p.cov.ucd <- lucd$plotdf %>%
    mutate(pos.idx = 1:nrow(.)) %>%
    ggplot(aes(x = pos.idx, y = log2(coverage + constant))) +
    geom_bar(stat = "identity", color = "black") +
    theme_classic() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank()) +
    scale_x_continuous(expand = c(0, 0)) +
    ylim(c(-1, 17)) +
    xlab("SNV position") +
    ylab("Log2 coverage")  
  
  p.freqcomb <- p.frac.vid / p.cov.vid / p.frac.ucd / p.cov.ucd +
    plot_layout(axes = 'collect_x',
                guides = 'collect')
  
  return(p.freqcomb)
  
}

```

Create plot for all samples:

```{r}
p.list <- list()
for (i in 1:length(all.samples)) {
  
  name <- all.samples[i]
  print(name)
  p.list[[name]] <- make_strainplots_panels(lucd = l.ucd.filt,
                                             lvid = l.vid.filt,
                                             sam = name,
                                             colors = strain.cols2,
                                             constant = sel.constant)
}

```

## KH14 {.tabset}

```{r, fig.width=14, fig.asp=0.5}
# print plots in report
for (i in 1:length(kh14.sams)) {
  
  cat("  \n")
  cat("  \n### ", glue("{kh14.sams[i]}"), " \n")
  cat("  \n")
  cat("<br>")
  
  figlab <- newlab(figlab, caption = "Base-level estimates")
  name <- kh14.sams[i]
  p.tmp <- p.list[[name]] +
    plot_annotation(title = figlab)
  print(p.tmp)
  
}

# Printing
for (i in 1:length(kh14.sams)) {
  
  name <- kh14.sams[i]
  outfile.base <- file.path(resultdir, "base-level-estimate-plots", glue("{name}-base-frac-cov"))
  print_baselevel_plot(p = p.list[[name]],
                       outfile.base, printpngs, printpdfs, set.dpi, 
                       height = ht.baselv, 
                       width = wt.baselv)
}

```

To investigate high VID frequencies inferred in UCD52 inoculum, export data tables.

```{r}
if (printtables) {

  # --- save UCD52mda mapped to VID fractions
  df <- l.vid.filt[["UCD52mda"]]$plotdf.fracs
  df.out <- file.path(procdir.7results, "UCD52mda-Rvid-strain-fracs.xlsx")
  write.xlsx(df, df.out)
  
  # --- save UCD52mda mapped to UCD fractions
  df <- l.ucd.filt[["UCD52mda"]]$plotdf.fracs
  df.out <- file.path(procdir.7results, "UCD52mda-Rucd-strain-fracs.xlsx")
  write.xlsx(df, df.out)
  
}

```

## Mixing {.tabset}

```{r, fig.width=14, fig.asp=0.5}
for (i in 1:length(mixing.sams)) {
  
  cat("  \n")
  cat("  \n### ", glue("{mixing.sams[i]}"), " \n")
  cat("  \n")
  cat("<br>")
  
  name <- mixing.sams[i]
  figlab <- newlab(figlab, caption = "Base-level estimates")
  p.tmp <- p.list[[name]] +
    plot_annotation(title = figlab)
  print(p.tmp)
  
}

# Printing
for (i in 1:length(mixing.sams)) {
  
  name <- mixing.sams[i]
  outfile.base <- file.path(resultdir, "base-level-estimate-plots", glue("{name}-base-frac-cov"))
  print_baselevel_plot(p = p.list[[name]],
                       outfile.base, printpngs, printpdfs, set.dpi, 
                       height = ht.baselv, 
                       width = wt.baselv)
}


```

# Base-level estimate plots -- all genome positions

- NOTE! Coverage panels are based on extrapolating coverage at each SNP.

Functions:
```{r}
make_strainplots_panels_Xgenome <- function(lucd, lvid, sam, 
                                            colors = strain.cols.Xgenome){
  
  # test
  # sam <- "KH14P14"
  # lvid <- l.vid.filt
  # lucd <- l.ucd.filt
  lvid <- lvid[[sam]]
  lucd <- lucd[[sam]]
  
  # Extend to all genomic positions
  p.frac.vid <- lvid$plotdf.fracs %>%
    filter(var.type.2 != "Low Coverage") %>%
    arrange(pos) %>%
    mutate( var.type.2 = factor(var.type.2, levels= names(colors)) ) %>%
    ggplot(aes(x = pos, y = frac, fill = var.type.2)) +
    geom_area() +
    theme_classic() +
    #theme(axis.text.x=element_blank(),
    #      axis.ticks.x=element_blank()) +
    scale_fill_manual(name = "",
                      values = colors) +
    xlab("Genome position") +
    ylab("Base fraction") +
    labs(title = sam,
         subtitle = "FL-RhCMV")

  p.cov.vid <- lvid$plotdf %>%
    filter(!lowcoverage) %>%
    arrange(pos) %>%
    ggplot(aes(x = pos, y = coverage)) +
    geom_area() +
    theme_classic() +
    #theme(axis.text.x=element_blank(),
    #      axis.ticks.x=element_blank()) +
    xlab("Genome position") +
    ylab("Total coverage")
  
  p.frac.ucd <- lucd$plotdf.fracs %>%
    filter(var.type.2 != "Low Coverage") %>%
    arrange(pos) %>%
    mutate( var.type.2 = factor(var.type.2, levels= names(colors)) ) %>%
    ggplot(aes(x = pos, y = frac, fill = var.type.2)) +
    geom_area() +
    theme_classic() +
    #theme(axis.text.x=element_blank(),
    #      axis.ticks.x=element_blank()) +
    scale_fill_manual(name = "",
                      values = colors) +
    xlab("Genome position") +
    ylab("Base fraction") +
    labs(subtitle = "RhCMV UCD52")
  
  p.cov.ucd <- lucd$plotdf %>%
    filter(!lowcoverage) %>%
    arrange(pos) %>%
    ggplot(aes(x = pos, y = coverage)) +
    geom_area() +
    theme_classic() +
    #theme(axis.text.x=element_blank(),
    #      axis.ticks.x=element_blank()) +
    xlab("Genome position") +
    ylab("Total coverage")
  
  p.freqcomb <- p.frac.vid / p.cov.vid / p.frac.ucd / p.cov.ucd +
    plot_layout(axes = 'collect_x',
                guides = 'collect')
  
  return(p.freqcomb)
  
}

```

Create plot for all samples:

```{r}
p.list2 <- list()
for (i in 1:length(all.samples)) {
  
  name <- all.samples[i]
  ptmp <- make_strainplots_panels_Xgenome(
    lucd = l.ucd.filt,
    lvid = l.vid.filt,
    sam = name
  )
  p.list2[[name]] <- ptmp
  
}

```

## KH14 {.tabset}

```{r, fig.width=14, fig.asp=0.5}
for (i in 1:length(kh14.sams)) {
  
  cat("  \n")
  cat("  \n### ", glue("{kh14.sams[i]}"), " \n")
  cat("  \n")
  cat("<br>")
  
  name <- kh14.sams[i]
  p <- p.list2[[name]]
  print(p)
  
}

# Printing
for (i in 1:length(kh14.sams)) {
  
  name <- kh14.sams[i]
  outfile.base <- file.path(resultdir, "base-level-estimate-plots", glue("{name}-base-frac-cov-Xgenome"))
  print_baselevel_plot(p = p.list2[[name]],
                       outfile.base, printpngs, printpdfs, set.dpi, 
                       height = ht.baselv, 
                       width = wt.baselv)
}


```

## Mixing {.tabset}

```{r, fig.width=14, fig.asp=0.5}
for (i in 1:length(mixing.sams)) {
  
  cat("  \n")
  cat("  \n### ", glue("{mixing.sams[i]}"), " \n")
  cat("  \n")
  cat("<br>")
  
  name <- mixing.sams[i]
  p <- p.list2[[name]]
  print(p)
  
}

# Printing
for (i in 1:length(mixing.sams)) {
  
  name <- mixing.sams[i]
  outfile.base <- file.path(resultdir, "base-level-estimate-plots", glue("{name}-base-frac-cov-Xgenome"))
  print_baselevel_plot(p = p.list2[[name]],
                       outfile.base, printpngs, printpdfs, set.dpi, 
                       height = ht.baselv, 
                       width = wt.baselv)
}

```

# Plot nucdiff SNPs along genome (barcode)

## R_VID_Q_UCD

```{r}
# --- pull positions in VID (FL)
vid.genome <- readDNAStringSet(vidref.fna)
pos.vec <- seq(1, length(vid.genome[[1]]))

# --- create logical vector of positions in UCD that differ
# --- from VID (nucdiff)
vid.genome.pos.vec <- pos.vec
vid.genome.nucdiff <- vid.genome.pos.vec %in% sSNPbases.vid$POS
plot.df <- data.frame(pos = vid.genome.pos.vec,
                      is.SNP = vid.genome.nucdiff,
                      fakeY = "genome_position")
p <- plot.df %>%
  ggplot(aes(x = pos, y = fakeY, 
             fill = is.SNP, color = is.SNP)) +
  geom_tile() +
  theme_void() +
  theme(axis.ticks = element_blank(),
        axis.text = element_blank(),
        axis.title = element_blank()) +
  scale_fill_manual(values = c("TRUE" = "black",
                               "FALSE" = "white"),
                    labels = c("FL-RhCMV", "RhCMV UCD52"),
                    name = "Strain") +
  scale_color_manual(values = c("TRUE" = "black",
                               "FALSE" = "white")) +
  guides(color = "none", fill = "none")

p

# Printing
outfile.base <- file.path(resultdir, "RvidQucd_nucdiff_genomewide_blackwhite")
print_baselevel_plot(p,
                     outfile.base, printpngs, printpdfs, set.dpi, 
                     height = ht.barcode, 
                     width = wt.barcode)

```

# Make filtering scheme panels

Functions:

```{r}
make_localdiffs_panel <- function(df.snpIndx.ref, max.localdiffs, pass.cols,
                                  hline.col = "darkgreen",
                                  bar.linewidth = .1){
  
  #df.snpIndx.ref <- df.snpdiffs.vid3
  
  plotdf.pass <- df.snpIndx.ref %>%
    dplyr::select(snp.pos, mean.ndiffs, pass.localdiffs) %>%
    arrange(snp.pos) %>%
    mutate(pos.idx = 1:nrow(.))
  
  # 1. Plot density of local differences
  p.localdiffs <- plotdf.pass %>%
    ggplot(aes(x = pos.idx, y = mean.ndiffs, 
               fill = pass.localdiffs, color = pass.localdiffs)) +
    geom_bar(stat = "identity", width = 1,
             linewidth = bar.linewidth) +
    xlab("SNV position") +
    ylab("Mean local\n differences") +
    geom_hline(yintercept = max.localdiffs, linetype = 2,
               color = hline.col) +
    scale_fill_manual(name = "Pass filter", values = pass.cols) +
    scale_color_manual(name = "Pass filter", values = pass.cols) +
    theme_classic() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank()) +
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_continuous(expand = c(0, 0)) 
  
  return(p.localdiffs)
  
}

make_inocov_panel <- function(df.snpIndx.ref, min.inocov, pass.cols,
                              hline.col = "darkgreen",
                              bar.linewidth = .1,
                              constant = 0.5){
  
  #df.snpIndx.ref <- df.snpdiffs.vid3
  
  plotdf.pass <- df.snpIndx.ref %>%
    dplyr::select(snp.pos, coverage, pass.inocov) %>%
    arrange(snp.pos) %>%
    mutate(pos.idx = 1:nrow(.))
  
  # 2. Plot alternate strain inoculum coverage
  p.inocov <- 
    plotdf.pass %>%
    mutate(log.coverage = log2(coverage + constant)) %>%
    ggplot(aes(x = pos.idx, y = log.coverage, 
               fill = pass.inocov, color = pass.inocov)) +
    geom_bar(stat = "identity", width = 1, linewidth = bar.linewidth) +
    xlab("SNV position") +
    ylab("Log2 alternate strain\n inoculum coverage") +
    geom_hline(yintercept = log2(min.inocov), linetype = 2, 
               color = hline.col) +
    scale_fill_manual(name = "Pass filter", values = pass.cols) +
    scale_color_manual(name = "Pass filter", values = pass.cols) +
    theme_classic() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank()) +
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_continuous(expand = c(0, 0)) 
  
  return(p.inocov)
  
}

make_inoAltfreq_panels <- function(df.snpIndx.ref, max.inoAltfreq, pass.cols,
                                   hline.col = "darkgreen",
                                   bar.linewidth = .1,
                                   returnVIDsample = T){
  
  #df.snpIndx.ref <- df.snpdiffs.vid3
  #bar.linewidth = .1
  #hline.col = "orange"
  
  plotdf.pass <- df.snpIndx.ref %>%
    dplyr::select(snp.pos, `965BACmda`, UCD52mda, fail.inoVID, fail.inoUCD) %>%
    arrange(snp.pos) %>%
    mutate(pos.idx = 1:nrow(.),
           pass.inoVID = !fail.inoVID,
           pass.inoUCD = !fail.inoUCD)
  
  # 3. Plot alternate strain inoculum frequency
  
  ## 965BACmda inoculum
  inoculum.sample <- "965BACmda"
  inoculum.sample.pretty <- "FL-RhCMV"
  strain.name <- "RhCMV UCD52"
  ylabtitle <- paste0(strain.name, " fraction in\n", inoculum.sample.pretty," inoculum")
  p.inoAltfreq1 <- plotdf.pass %>%
    filter(!is.nan(`965BACmda`)) %>%
    ggplot(aes(x = pos.idx, y = `965BACmda`, 
               fill = pass.inoVID, color = pass.inoVID)) +
    geom_bar(stat = "identity", width = 1,  linewidth = bar.linewidth) +
    xlab("SNV position") +
    ylab(ylabtitle) +
    geom_hline(yintercept = max.inoAltfreq, linetype = 2,
               color = hline.col) +
    scale_fill_manual(name = "Pass filter", values = pass.cols) +
    scale_color_manual(name = "Pass filter", values = pass.cols) +
    theme_classic() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank()) +
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_continuous(limits = c(0,1), expand = c(0, 0)) 
  
  ## UCD52mda inoculum
  inoculum.sample <- "UCD52mda" 
  inoculum.sample.pretty <- "RhCMV UCD52"
  strain.name <- "FL-RhCMV"
  ylabtitle <- paste0(strain.name, " fraction in\n ", inoculum.sample.pretty," inoculum")
  p.inoAltfreq2 <- plotdf.pass %>%
    filter(!is.nan(`UCD52mda`)) %>%
    ggplot(aes(x = pos.idx, y = `UCD52mda`, 
               fill = pass.inoUCD, color = pass.inoUCD)) +
    geom_bar(stat = "identity", width = 1, linewidth = bar.linewidth) +
    xlab("SNV position") +
    ylab(ylabtitle) +
    geom_hline(yintercept = max.inoAltfreq, linetype = 2,
               color = hline.col) +
    scale_fill_manual(name = "Pass filter", values = pass.cols) +
    scale_color_manual(name = "Pass filter", values = pass.cols) +
    theme_classic() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank()) +
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_continuous(limits = c(0,1), expand = c(0, 0)) 
  
  if(returnVIDsample){
    p <-  p.inoAltfreq1
  }else{
    p <-  p.inoAltfreq2
  }
  
  return(p)
  
}

make_filterTile_panel <- function(df.snpIndx.ref, pass.cols){
  
  #df.snpIndx.ref <- df.snpdiffs.vid3
  
  plotdf.pass <- df.snpIndx.ref %>%
    dplyr::select(snp.pos, pass.localdiffs, pass.inocov, pass.inoAltfreq) %>%
    arrange(snp.pos) %>%
    mutate(pos.idx = 1:nrow(.))
  
  # 4. Plot passed filter tiles
  tmpdf <- plotdf.pass %>%
    pivot_longer(cols = c(pass.localdiffs, pass.inocov, pass.inoAltfreq),
                 names_to = "pass.type",
                 values_to = "pass.value") %>%
    mutate(pass.value = factor(pass.value))
  
  namesdf <- data.frame(pass.type = c("pass.localdiffs","pass.inocov","pass.inoAltfreq"),
             pass.type.pretty = c("Local differences (I)",
                                  "Alt-strain inoculum coverage (J)",
                                  "Alt-strain inoculum fraction (K, L)")
             )
  p.pass <- tmpdf %>%
    left_join(namesdf, by = c("pass.type")) %>%
    mutate(pass.type.pretty = factor(pass.type.pretty, levels = rev(namesdf$pass.type.pretty))) %>% 
    ggplot(aes(x = pos.idx, y = pass.type.pretty, fill = pass.value, color = pass.value)) +
    geom_tile() +
    scale_fill_manual(name = "Pass filter", values = pass.cols) +
    scale_color_manual(name = "Pass filter", values = pass.cols) +
    xlab("SNV position") +
    ylab("") +
    theme_classic() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank(),
          axis.text.y = element_text(hjust = 0.5),
          axis.ticks.y=element_blank()) +
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_discrete(expand = c(0, 0)) 
  
  return(p.pass)
  
}

make_prefiltFrac_panel <- function(sam, l.ref, colors = strain.cols.filt,
                                   bar.linewidth = .1){
  
  strain.cols2.labels <- names(strain.cols.filt)
  
  # testing
  #sam <- "KH14heart" #sam <- "UCD52mda"
  #l.ref <- l.vid
  
  lref <- l.ref[[sam]]
  plotdf.fracs <- lref$plotdf.fracs %>%
    arrange(pos) %>%
    mutate(var.type.3 = ifelse(is.nan(frac), "Zero Coverage", var.type.2),
           var.type.3 = factor(var.type.3, levels = strain.cols2.labels),
           frac.2 = ifelse(is.nan(frac), 1/3, frac),
           pos.idx = rep(1:length(unique(pos)), each = 3)) 
  
  # 5. Plot pre-filtered strain frequencies
  p.frac <- plotdf.fracs %>%
    ggplot(aes(x = pos.idx, y = frac.2, 
               fill = var.type.3, color = var.type.3)) +
    geom_bar(stat = "identity", position = "stack", width = 1,
             linewidth = bar.linewidth) +
    scale_fill_manual(name = "Base strain match",
                      values = colors) +
    scale_color_manual(name = "Base strain match",
                       values = colors) +
    xlab("SNV position") +
    ylab("Base fraction") +
    theme_classic() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank()) +
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_continuous(limits = c(0,1), expand = c(0, 0)) 
  
  #p.frac
  
  return(p.frac)
  
}

make_postfiltFrac_panel <- function(sam, l.ref, df.snpIndx.ref, 
                                    colors = strain.cols3,
                                    bar.linewidth = .1){
  
  strain.cols3.labels <- names(strain.cols3)
  
  # testing
  # sam <- "965BACmda" 
  # sam <- "UCD52mda"
  # l.ref <- l.vid
  # df.snpIndx.ref <- df.snpdiffs.vid3
  
  # get pass info
  df.pass.indx <- df.snpIndx.ref %>%
    dplyr::select(snp.pos, pass.localdiffs, pass.inocov, pass.inoAltfreq) %>%
    mutate(pass.position = ifelse(pass.localdiffs == FALSE |
                                    pass.inocov == FALSE |
                                    pass.inoAltfreq == FALSE, FALSE, TRUE)) 
  
  lref <- l.ref[[sam]]
  plotdf.fracs <- lref$plotdf.fracs %>%
    arrange(pos) %>%
    mutate(pos.idx = rep(1:length(unique(pos)), each = 3)) %>%
    left_join(df.pass.indx, by = c("pos"="snp.pos")) %>%
    mutate(var.type.3 = ifelse(pass.position == FALSE, "Filtered Position", var.type.2),
           var.type.3 = factor(var.type.3, levels = strain.cols3.labels),
           frac.2 = ifelse(pass.position == FALSE, 1/3, frac)) 
  
  # 6. Plot post-filtered strain frequencies
  p.frac <- plotdf.fracs %>%
    ggplot(aes(x = pos.idx, y = frac.2, 
               fill = var.type.3, color = var.type.3)) +
    geom_bar(stat = "identity", position = "stack", width = 1,
             linewidth = bar.linewidth) +
    scale_fill_manual(name = "Base strain match",
                      values = colors) +
    scale_color_manual(name = "Base strain match",
                      values = colors) +
    xlab("SNV position") +
    ylab("Base fraction") +
    theme_classic() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank()) +
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_continuous(limits = c(0,1), expand = c(0, 0)) 
  
  return(p.frac)
  
}

make_coverage_panel <- function(sam, l.ref, 
                                bar.linewidth = .1,
                                constant = 0.5){
  
  # testing
  #sam <- "965BACmda" #sam <- "UCD52mda"
  # l.ref <- l.vid
  
  lref <- l.ref[[sam]]
  plotdf <- lref$plotdf %>%
    mutate(pos.idx = 1:nrow(.))
  
  # 6. Plot coverage
  p.cov <- 
    plotdf %>%
      mutate(log.coverage = log2(coverage + constant)) %>%
    ggplot(aes(x = pos.idx, y = log.coverage)) +
    geom_bar(stat = "identity", width = 1, 
             linewidth = bar.linewidth,
             color = "black", fill = "black") +
    xlab("SNV position") +
    ylab("Log2 coverage") +
    theme_classic() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank()) +
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_continuous(expand = c(0, 0)) 
  
  return(p.cov)
  
}

```

Create filtering scheme panels using VID as the reference

```{r}
# Retain strain SNP positions with **fewer than 5** mean local differences in the 100bp windows that intersect each SNP
max.localdiffs <- 5

# Retain strain SNP positions with **greater than 250** reads
min.inocov <- 250

# Retain strain SNP positions with less than 1% alternate strain frequency in the sequenced inoculum samples
max.inoAltfreq <- 0.01 # 1%

## VID ##---------------------------------------------------##

reference <- "VID"
df.snpIndx.ref <- df.snpdiffs.vid3
l.ref <- l.vid


# Filter panels ##---------------------------------##

# make figures
p.localdiffs <- make_localdiffs_panel(df.snpIndx.ref, max.localdiffs, pass.cols)
p.inocov <- make_inocov_panel(df.snpIndx.ref, min.inocov, pass.cols,
                              constant = sel.constant)
p.inoAltfreq1 <- make_inoAltfreq_panels(df.snpIndx.ref, max.inoAltfreq, pass.cols,
                                       returnVIDsample = T)
p.inoAltfreq2 <- make_inoAltfreq_panels(df.snpIndx.ref, max.inoAltfreq, pass.cols,
                                       returnVIDsample = F)
p.filterTile <- make_filterTile_panel(df.snpIndx.ref, pass.cols)

# store panels
p.l <- list(localdiffs = p.localdiffs,
                     inocov = p.inocov,
                     inoAltfreq1 = p.inoAltfreq1,
                     inoAltfreq2 = p.inoAltfreq2,
                     filterTile = p.filterTile)
# add fonts
p.l1 <- p.l %>%
  map(~.x +
        theme(text = element_text(family = "Arimo")))
# remove legends
pNoLegends.l1 <- p.l1 %>%
  map(~.x +
        guides(color = "none", fill = "none"))


# Inoculum sample panels ##---------------------------------##

# make figures
sam <- "965BACmda"
p.prefilt1 <- make_prefiltFrac_panel(sam, l.ref)
p.postfilt1 <- make_postfiltFrac_panel(sam, l.ref, df.snpIndx.ref)
p.cov1 <- make_coverage_panel(sam, l.ref, constant = sel.constant)
sam <- "UCD52mda"
p.prefilt2 <- make_prefiltFrac_panel(sam, l.ref)
p.postfilt2 <- make_postfiltFrac_panel(sam, l.ref, df.snpIndx.ref)
p.cov2 <- make_coverage_panel(sam, l.ref, constant = sel.constant)

# store panels
ps.l <- list(prefilt1 = p.prefilt1,
             postfilt1 = p.postfilt1,
             cov1 = p.cov1,
             prefilt2 = p.prefilt2,
             postfilt2 = p.postfilt2,
             cov2 = p.cov2)
# add fonts
p.l2 <- ps.l %>%
  map(~.x +
        theme(text = element_text(family = "Arimo")))
# remove legends
pNoLegends.l2 <- p.l2 %>%
  map(~.x +
        guides(color = "none", fill = "none"))


# Patchwork the panels ##---------------------------------##

# Panels B-H
panels.bh <- pNoLegends.l2$prefilt1 + pNoLegends.l2$prefilt2 +
  plot_spacer() + free(pNoLegends.l1$filterTile, type = "space", side = "l") +
  pNoLegends.l2$postfilt1 + pNoLegends.l2$postfilt2 +
  pNoLegends.l2$cov1 + pNoLegends.l2$cov2 +
  plot_layout(ncol = 2)

# Panels I-L
panels.il <- pNoLegends.l1$localdiffs +
  pNoLegends.l1$inocov +
  pNoLegends.l1$inoAltfreq1 +
  pNoLegends.l1$inoAltfreq2 +
  plot_layout(ncol = 1)

# Pass filter colors
leg <-  ggpubr::as_ggplot(ggpubr::get_legend(p.localdiffs))

# Pre-filter colors
leg.prefilt <-  ggpubr::as_ggplot(ggpubr::get_legend(p.prefilt2))

# Post-filter colors
leg.postfilt <-  ggpubr::as_ggplot(ggpubr::get_legend(p.postfilt1))

```

Show filtering scheme panels

```{r, fig.width=14, fig.asp=0.5}

figlab <- newlab(figlab, caption = "Panels B-H")
panels.bh +
  plot_annotation(title = figlab)

figlab <- newlab(figlab, caption = "Panels I-L")
panels.il +
  plot_annotation(title = figlab)

figlab <- newlab(figlab, caption = "Pass legend")
leg +
  plot_annotation(title = figlab)

figlab <- newlab(figlab, caption = "Pre-filter legend")
leg.prefilt +
  plot_annotation(title = figlab)

figlab <- newlab(figlab, caption = "Post-filter legend")  
leg.postfilt +
  plot_annotation(title = figlab)


# Printing

# - Panels B-H
out.width <- wt.filtpanel * 2 # cols
out.ht <- ht.filtpanel * 4 # rows
outfile.base <- file.path(resultdir, "filter-panels", paste0("PanelsB-H_ref", reference))
print_baselevel_plot(p = panels.bh,
                     outfile.base, printpngs, printpdfs, set.dpi, 
                     height = out.ht, 
                     width = out.width)

# - Panels I-L
out.width <- wt.filtpanel * 1 # cols
out.ht <- ht.filtpanel  * 4 # rows
outfile.base <- file.path(resultdir, "filter-panels", paste0("PanelsI-L_ref", reference))
print_baselevel_plot(p = panels.il,
                     outfile.base, printpngs, printpdfs, set.dpi, 
                     height = out.ht, 
                     width = out.width)

# - Legend: Pass filter colors
outfile.base <- file.path(resultdir, "filter-panels", paste0("legend-PassFilter_ref", reference))
print_baselevel_plot(p = leg,
                     outfile.base, printpngs, printpdfs, set.dpi, 
                     height = ht.legend, 
                     width = wt.legend)

# - Legend: Pre-filter colors
outfile.base <- file.path(resultdir, "filter-panels", paste0("legend-PreFilt_ref", reference))
print_baselevel_plot(p = leg.prefilt,
                     outfile.base, printpngs, printpdfs, set.dpi, 
                     height = ht.legend, 
                     width = wt.legend)
  
# - Legend: Post-filter colors
outfile.base <- file.path(resultdir, "filter-panels", paste0("legend-PostFilt_ref", reference))
print_baselevel_plot(p = leg.postfilt,
                     outfile.base, printpngs, printpdfs, set.dpi, 
                     height = ht.legend, 
                     width = wt.legend)


```



# Session information

```{r, results='hold'}
sessionInfo()

print(glue("Start Time: ",startdt))
print(glue("End Time: ",date()))

```
